File: 3753.total-waviness-of-numbers-in-range-ii.java
#
# @lc app=leetcode id=3753 lang=java
#
# [3753] Total Waviness of Numbers in Range II
#

# @lc code=start
class Solution {
    public long totalWaviness(long num1, long num2) {
        return calc(num2) - calc(num1 - 1);
    }

    private long calc(long n) {
        if (n <= 0) return 0L;
        String str = String.valueOf(n);
        int L = str.length();
        int[] digits = new int[L];
        for (int i = 0; i < L; i++) {
            digits[i] = str.charAt(i) - '0';
        }
        long[][][][] cntMemo = new long[L + 1][2][11][11];
        long[][][][] sumMemo = new long[L + 1][2][11][11];
        for (int i = 0; i <= L; i++) {
            for (int j = 0; j < 2; j++) {
                for (int k = 0; k < 11; k++) {
                    for (int m = 0; m < 11; m++) {
                        cntMemo[i][j][k][m] = -1L;
                        sumMemo[i][j][k][m] = -1L;
                    }
                }
            }
        }
        long[] res = solve(0, 1, 0, 0, cntMemo, sumMemo, digits, L);
        return res[1];
    }

    private long[] solve(int idx, int tight, int p2, int p1, long[][][][] cntMemo, long[][][][] sumMemo, int[] digits, int L) {
        if (idx == L) {
            return new long[]{p1 != 0 ? 1L : 0L, 0L};
        }
        if (cntMemo[idx][tight][p2][p1] != -1L) {
            return new long[]{cntMemo[idx][tight][p2][p1], sumMemo[idx][tight][p2][p1]};
        }
        long tot_cnt = 0L;
        long tot_sum = 0L;
        int up = tight == 1 ? digits[idx] : 9;
        for (int d = 0; d <= up; d++) {
            int ntight = (tight == 1 && d == up) ? 1 : 0;
            int np2, np1;
            if (p1 == 0 && d == 0) {
                np2 = 0;
                np1 = 0;
            } else {
                np1 = d + 1;
                np2 = p1;
            }
            long[] res = solve(idx + 1, ntight, np2, np1, cntMemo, sumMemo, digits, L);
            long this_cnt = res[0];
            long this_sumw = res[1];
            long contrib = 0L;
            if (p1 != 0 && p2 != 0) {
                int leftv = p2 - 1;
                int centerv = p1 - 1;
                int rightv = d;
                if ((centerv > leftv && centerv > rightv) || (centerv < leftv && centerv < rightv)) {
                    contrib = this_cnt;
                }
            }
            tot_cnt += this_cnt;
            tot_sum += this_sumw + contrib;
        }
        cntMemo[idx][tight][p2][p1] = tot_cnt;
        sumMemo[idx][tight][p2][p1] = tot_sum;
        return new long[]{tot_cnt, tot_sum};
    }
}
# @lc code=end

File: 3753.total-waviness-of-numbers-in-range-ii.go
#
# @lc app=leetcode id=3753 lang=golang
#
# [3753] Total Waviness of Numbers in Range II
#

# @lc code=start
func totalWaviness(num1 int64, num2 int64) int64 {
	calc := func(n int64) int64 {
		if n < 1 {
			return 0
		}

		// Extract digits LSB first
		var digs []int
		tmp := n
		for tmp > 0 {
			digs = append(digs, int(tmp%10))
			tmp /= 10
		}
		// Reverse to MSB first
		for i, j := 0, len(digs)-1; i < j; i, j = i+1, j-1 {
			digs[i], digs[j] = digs[j], digs[i]
		}
		L := len(digs)

		const MAXS = 70010
		memo_wav := make([]int64, MAXS)
		memo_cnt := make([]int64, MAXS)
		for i := 0; i < MAXS; i++ {
			memo_wav[i] = -1
			memo_cnt[i] = -1
		}

		stride_slen := 1
		stride_p1 := 17
		stride_p2 := 11 * stride_p1 // 187
		stride_tight := 11 * stride_p2 // 2057
		stride_pos := 2 * stride_tight // 4114

		getIndex := func(pos, tight, p2, p1, slen int) int {
			return pos*stride_pos + tight*stride_tight + p2*stride_p2 + p1*stride_p1 + slen
		}

		var dfs func(pos, tight, p2, p1, slen int) (int64, int64)
		dfs = func(pos, tight, p2, p1, slen int) (int64, int64) {
			if pos == L {
				return 0, 1
			}

			idx := getIndex(pos, tight, p2, p1, slen)
			if memo_wav[idx] != -1 {
				return memo_wav[idx], memo_cnt[idx]
			}

			var aw, ac int64 = 0, 0
			up := 9
			if tight == 1 {
				up = digs[pos]
			}

			for dd := 0; dd <= up; dd++ {
				nt := 0
				if tight == 1 && dd == up {
					nt = 1
				}

				contrib := int64(0)
				np2 := p2
				np1 := p1
				nslen := slen

				if !(slen == 0 && dd == 0) {
					oldslen := slen
					if oldslen >= 2 {
						if (p1 > p2 && p1 > dd) || (p1 < p2 && p1 < dd) {
							contrib = 1
						}
					}
					nslen = oldslen + 1
					np1 = dd
					if oldslen >= 1 {
						np2 = p1
					} else {
						np2 = 10
					}
				}

				sw, sc := dfs(pos+1, nt, np2, np1, nslen)
				aw += contrib * sc + sw
				ac += sc
			}

			memo_wav[idx] = aw
			memo_cnt[idx] = ac
			return aw, ac
		}

		res, _ := dfs(0, 1, 10, 10, 0)
		return res
	}

	return calc(num2) - calc(num1-1)
}
# @lc code=end